File: SeidorSmallWebApi/UDO/SEI_ConfOrders.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Threading.Tasks;
using System.Xml.Linq;
using CrystalDecisions.CrystalReports.Engine;
using Newtonsoft.Json.Linq;
using SAPbobsCOM;
using SAPbouiCOM;
using SEI.MonitorDeComandes.SEI_AddonEnum;
using SEIDOR_SLayer;

public class SEI_ConfOrders : SEI_Form
{
    #region ATTRIBUTES

    #endregion

    #region CONSTANTS
    private struct FormControls
    {
        // Define your control IDs here
    }
    #endregion

    #region CONSTRUCTOR
    public SEI_ConfOrders(SEI_Addon parentAddon)
        : base(parentAddon, enSBO_LoadFormTypes.XmlFile, enAddonFormType.f_ConfOrders, enAddonMenus.ConfOrders)
    {
        this.Initialize();
    }

    private void Initialize()
    {
        try
        {
            this.Form.Visible = true;
        }
        catch (Exception ex)
        {
            SBO_Application.StatusBar.SetText("Error loading 'Confirmation Orders' screen", BoMessageTime.bmt_Short, BoStatusBarMessageType.smt_Error);
            this.Form.Close();
        }
    }
    #endregion

    #region EVENT FUNCTIONS
    public override void HANDLE_DATA_EVENT(ref BusinessObjectInfo BusinessObjectInfo, ref bool BubbleEvent)
    {
        // Handle business object data events
    }

    public override void HANDLE_FORM_EVENTS(string FormUID, ref ItemEvent pVal, ref bool BubbleEvent)
    {
        try
        {
            if (!string.IsNullOrEmpty(pVal.ItemUID.Trim()))
            {
                // Handle control or form item events
                switch (pVal.ItemUID)
                {
                    //case FormControls.btnSend:
                    //    HandleBtnSend(ref pVal, ref BubbleEvent);
                    //    break;
                }
            }
        }
        catch (Exception ex)
        {
            SBO_Application.StatusBar.SetText($"Screen error: {ex.Message}", BoMessageTime.bmt_Short, BoStatusBarMessageType.smt_Error);
        }
    }

    public override void HANDLE_MENU_EVENTS(ref MenuEvent pVal, ref bool BubbleEvent)
    {
        // Handle menu events
    }

    public override void HANDLE_PRINT_EVENT(ref PrintEventInfo eventInfo, ref bool BubbleEvent)
    {
        // Handle print events
    }

    public override void HANDLE_REPORT_DATA_EVENT(ref ReportDataInfo eventInfo, ref bool BubbleEvent)
    {
        // Handle report data events
    }

    public override void HANDLE_RIGHTCLICK_EVENT(ref IContextMenuInfo eventInfo, ref bool BubbleEvent)
    {
        // Handle right-click events
    }

    public override void HANDLE_LUPA(ref string CampoLupa, ref List<ArrayList> aRetorno)
    {
        // Handle "lupa" (choose from list) events
    }

    public override void HANDLE_LAYOUTKEY_EVENTS(ref LayoutKeyInfo EventInfo, ref bool BubbleEvent)
    {
        // Handle layout key events
    }

    public override void HANDLE_PROGRESSBAR_EVENTS(ref IProgressBarEvent pVal, ref bool BubbleEvent)
    {
        // Handle progress bar events
    }

    public override void HANDLE_STATUSBAR_EVENTS(ref string Text, ref BoStatusBarMessageType MessageType)
    {
        // Handle status bar events
    }

    public override void HANDLE_WIDGET_EVENTS(ref WidgetData pWidgetData, ref bool BubbleEvent)
    {
        // Handle widget events
    }
    #endregion

    #region FORM FUNCTIONS

    #endregion

    #region FUNCTIONS
    #region INITIALIZERS
    // Add initializer methods here

    #endregion

    #region HANDLERS
    // Add event handler methods here

    #endregion

    #region GENERAL FUNCTIONS
    // Helper method to load UDO data
    public void LoadUDOData(string docEntry)
    {
        try
        {
            // Get UDO data using Service Layer connection
            var udo = SBO_Company.GetBusinessObject(BoObjectTypes.BoRecordset) as Recordset;
            string query = $@"
                SELECT * FROM [@CONF_ORDERS] 
                WHERE U_DocEntry = '{docEntry}'";

            udo.DoQuery(query);

            if (!udo.EoF)
            {
                // Populate header data
                string status = udo.Fields.Item("U_Status").Value.ToString();
                DateTime? confDate = null;
                if (udo.Fields.Item("U_ConfDate").Value != null)
                {
                    confDate = DateTime.Parse(udo.Fields.Item("U_ConfDate").Value.ToString());
                }

                // Update form fields
                // Form.Items.Item("txtDocEntry").Specific.Value = docEntry;
                // Form.Items.Item("txtStatus").Specific.Value = status;
                // if (confDate.HasValue)
                // {
                //     Form.Items.Item("dtConfDate").Specific.Value = confDate.Value;
                // }

                // Load detail records
                LoadDetailLines(docEntry);
            }
        }
        catch (Exception ex)
        {

File: MonitorDeComandes/Forms Add-on/SEI_ConfOrders.vb
Imports System.Collections
Imports System.Threading.Tasks
Imports Newtonsoft.Json.Linq
Imports SAPbobsCOM
Imports SAPbouiCOM
Imports SEI.MonitorDeComandes.SEI_AddonEnum
Imports SEIDOR_SLayer

Public Class SEI_ConfOrders
    Inherits SEI_Form

#Region "ATRIBUTOS"

#End Region

#Region "CONSTANTES"
    Private Structure FormControls


    End Structure
#End Region

#Region "CONSTRUCTOR"
    Public Sub New(ByRef parentaddon As SEI_Addon)
        MyBase.New(parentaddon, enSBO_LoadFormTypes.XmlFile, enAddonFormType.f_ConfOrders, enAddonMenus.ConfOrders, "CONFORDERS")
        Me.Initialize()
    End Sub

    Private Sub Initialize()
        Try
            ConfigurarMatrix()
            Me.Form.Visible = True
        Catch ex As Exception
            SBO_Application.StatusBar.SetText("Error al cargar la pantalla 'Confirmation Orders'", SAPbouiCOM.BoMessageTime.bmt_Short, SAPbouiCOM.BoStatusBarMessageType.smt_Error)
            Me.Form.Close()
        End Try
    End Sub

#End Region

#Region "FUNCIONES EVENTOS"
    Public Overrides Sub HANDLE_DATA_EVENT(ByRef BusinessObjectInfo As SAPbouiCOM.BusinessObjectInfo, ByRef BubbleEvent As Boolean)
    End Sub

    Public Overrides Sub HANDLE_FORM_EVENTS(FormUID As String, ByRef pVal As ItemEvent, ByRef BubbleEvent As Boolean)


        If Trim(pVal.ItemUID) <> "" Then
            Select Case pVal.ItemUID
            End Select

        Else
            '
            'Eventos de Formulario
            '
            Select Case pVal.EventType
                Case SAPbouiCOM.BoEventTypes.et_FORM_ACTIVATE
                    '
                Case SAPbouiCOM.BoEventTypes.et_FORM_LOAD
                    ' AFTER
                    ' Se captura al inicializar la clase Initialize()
                Case SAPbouiCOM.BoEventTypes.et_FORM_CLOSE

                Case SAPbouiCOM.BoEventTypes.et_FORM_UNLOAD

                Case SAPbouiCOM.BoEventTypes.et_FORM_RESIZE

            End Select
        End If
    End Sub

    Public Overrides Sub HANDLE_MENU_EVENTS(ByRef pVal As SAPbouiCOM.MenuEvent, ByRef BubbleEvent As Boolean)
    End Sub

    Public Overrides Sub HANDLE_PRINT_EVENT(ByRef eventInfo As SAPbouiCOM.PrintEventInfo, ByRef BubbleEvent As Boolean)
    End Sub

    Public Overrides Sub HANDLE_REPORT_DATA_EVENT(ByRef eventInfo As SAPbouiCOM.ReportDataInfo, ByRef BubbleEvent As Boolean)
    End Sub

    Public Overrides Sub HANDLE_RIGHTCLICK_EVENT(ByRef eventInfo As SAPbouiCOM.IContextMenuInfo, ByRef BubbleEvent As Boolean)
    End Sub

    Public Overrides Sub HANDLE_LUPA(ByRef CampoLupa As String, ByRef aRetorno As System.Collections.Generic.List(Of System.Collections.ArrayList))
    End Sub

    Public Overrides Sub HANDLE_LAYOUTKEY_EVENTS(ByRef EventInfo As SAPbouiCOM.LayoutKeyInfo, ByRef BubbleEvent As Boolean)
    End Sub

    Public Overrides Sub HANDLE_PROGRESSBAR_EVENTS(ByRef pVal As SAPbouiCOM.IProgressBarEvent, ByRef BubbleEvent As Boolean)
    End Sub

    Public Overrides Sub HANDLE_STATUSBAR_EVENTS(ByRef Text As String, ByRef MessageType As SAPbouiCOM.BoStatusBarMessageType)
    End Sub

    Public Overrides Sub HANDLE_WIDGET_EVENTS(ByRef pWidgetData As SAPbouiCOM.WidgetData, ByRef BubbleEvent As Boolean)
    End Sub
    Private Sub HandleChooseFromListEvent(ByRef pVal As SAPbouiCOM.ItemEvent, ByRef BubbleEvent As Boolean)
    End Sub
#End Region

#Region "FUNCIONES FORM"

    Private Sub ConfigurarMatrix()
        Try
            Dim oMatrix As SAPbouiCOM.Matrix = CType(Form.Items.Item("Item_5").Specific, SAPbouiCOM.Matrix)

            If oMatrix IsNot Nothing Then
                With oMatrix
                    .Columns.Item("#").Width = 22
                    .Columns.Item("C_0_3").Width = 130
                    .Columns.Item("C_0_1").Width = 130
                    .Columns.Item("C_0_2").Width = 130
                    .Columns.Item("C_0_4").Width = 130
                End With
            End If

        Catch ex As Exception
            SBO_Application.StatusBar.SetText($"Error al configurar Matrix: {ex.Message}",
            SAPbouiCOM.BoMessageTime.bmt_Short,
            SAPbouiCOM.BoStatusBarMessageType.smt_Error)
        End Try
    End Sub

#End Region

#Region "FUNCIONES GENERALES"


#End Region
End Class